### PR TITLE
[Tooling] Automate moving the PRs to next milestone during code-freeze

### PR DESCRIPTION
**Pull request by AliSoftware**
## Description

This automates the migration of PRs to the next milestone during code freeze.

Right now, at the start of the release scenario in MC for version `x.y`, one of the instructions is to move all still-opened PRs targeting the milestone `x.y` to milestone `x.y+1`. This PR now automates that step, not only updating the milestone of those still-opened PRs but also commenting on them to let the PR author know, and reporting the list of which PRs have been updated.

This change will be accompanied by a corresponding change in the Release Scenario in MC to remove the manual task (internal ref: D167967-code)

## Testing Instructions

Won't really be able to test this until next real code freeze.

### CODE REVIEW

File: fastlane/Fastfile
Comment (dangermattic) [inline fastlane/Fastfile:156]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-android/blob/14b7e20c8217e600dd05302a97a4056aa980069a/fastlane/Fastfile#L156"></span>Lint/UselessAssignment: Useless assignment to variable - <code>moved_prs</code>.</td>
    </tr>
  </tbody>
</table>

File: fastlane/Fastfile
Comment (MiSikora) [inline fastlane/Fastfile:346]: I'm not sure if this covers that but we start our prod releases at 1% on Fridays, then on Mondays we bump to 10% and rollout from there.

**Comment by MiSikora** [comment]
> This change will be accompanied by a corresponding change in the Release Scenario in MC to remove the manual task.

Will there still be a step to verify hanging PRs first? Sometimes we want to add changes on the day of the release and it would be good to keep this in the MC flow.

File: fastlane/Fastfile
Comment (AliSoftware) [reply fastlane/Fastfile:346]: @MiSikora  Thanks for asking. I saw that from the MC scenario, but is 1% really ever in effect in practice though?

Because what I read from the scenario is that:
 - You **_submit for review_** with a rollout of 1% on Friday…
 - But once it's approved by Google, you update that rollout value to 10% on Monday on the moment that you release the approved version?

Or am I misreading the current scenario? Maybe the current scenario is missing a task about «Click "Publish N changes» at the start of Monday, before the currently-first task of Monday that is about updating the rollout to 10%, meaning that you start early Monday morning by clicking the button to publish to 1% and only later on that same Monday you update the rollout?

**Comment by AliSoftware** [comment]
> > This change will be accompanied by a corresponding change in the Release Scenario in MC to remove the manual task.
> 
> Will there still be a step to verify hanging PRs first? Sometimes we want to add changes on the day of the beta release and it would be good to keep this in the MC flow.

I'll keep a manual task about checking those then 👍

File: fastlane/Fastfile
Comment (MiSikora) [reply fastlane/Fastfile:346]: I believe we've always submitted for publication at 1% on Fridays. At least, that's what I’ve been doing, and from my memory, others have as well, since I’ve received emails about live rollouts over the weekend. I think publishing at 1% over the weekend isn’t a bad idea. The volume is extremely low, so the risk is minimal, and it gives us a chance to catch any crashes that might occur during that time.

That said, I’m open to changing our publishing workflow to what you described, as it seems like that’s how it should have been. 😓

Comment (AliSoftware) [reply fastlane/Fastfile:346]: I'm not sure I'm following. When you submit on Friday, you only submit to Google for Review.

The way "Managed Publishing" works in Google Play is that you submit for review, then Google reviews and (hopefully) approve your submission… but doesn't publish it yet, and it's only when you go to the "Publishing Overview" page after they approved it and explicitly click on the "Publish N changes" button there that the rollout actually starts.

The rollout does not start immediately after Google approves the submission, but only when you explicitly click on the "Publish Changes" button.

Hence why I'm saying that having the rollout set to 1% during the weekend does not do anything (unless Google speed-approves your submission on Friday before your Friday EOD and you decide to click the "Publish Changes" on Friday before the weekend instead of waiting for Monday to do that… but I don't think you're doing that either)

Comment (MiSikora) [reply fastlane/Fastfile:346]: I'm saying I, and from what I can tell we as a team, haven't been using Managed Publishing when submitting on Fridays.

Comment (AliSoftware) [reply fastlane/Fastfile:346]: You can't choose to use Managed Publishing or not on a case by case basis.

Managed Publishing is turned on for PocketCasts in Play Store, and that affects all submissions (and you can't turn it off unless you ask Systems as you need special permissions for that iinm, and that would affect all the builds anyway, it's an all-or-nothing setting in Play Store Console)
image

So given Managed Publishing is turned on, this has always worked like described in [this Google help page](https://support.google.com/googleplay/android-developer/answer/9859654?hl=en).

When you submit for review on Friday, the submitted build will be in "Changes submitted for Review" section of the "Publishing Overview" page, and Google will review them at some point over the weekend, but the rollout won't start until you explicitly click "Publish N changes" for those approved changes on Monday. So nothing is ever rolled out during the weekend, because Managed Publishing is on in our Play Store Console settings for the whole project.

Comment (MiSikora) [reply fastlane/Fastfile:346]: Sorry, but we're going in circles. I definitely can turn managed publishing on and off, and I'm sure that others can as well since we've been doing it in the past. Just from the rollout history:

* `7.77` - Saturday, November 23rd at 1%
* `7.76` - Saturday, November 9th at 1%
* `7.75` - Saturday, October 26th at 1%
* `7.74` - Monday, October 14th at 1%
* `7.73` - Saturday, September 28th at 1%

Comment (AliSoftware) [reply fastlane/Fastfile:346]: Wait if you've actively been turning Managed Publishing on and off on the PlayStore (turning it off on Fridays, and remembering to turn it back on on Monday, manually each time?!) then I'm confused because it's documented nowhere in the MC scenario that you'd have to do that manually on each Friday and Monday 🤔

Comment (MiSikora) [reply fastlane/Fastfile:346]: I haven't. It's been off by default when I've been publishing. I remember turning it off once when we had to coordinate a release of one feature with a partner.

Comment (AliSoftware) [reply fastlane/Fastfile:346]: Ok. I think you're the only Android product in a8c that is not using Managed Publishing all the time then 😅 

I'm still confused why it's not explicitly documented in the release scenario that one would have to decide to turn it on or off; especially given right now it's turned ON (just like for all our other a8c products in Google Play) and I'd have assumed you don't keep changing that setting constantly.

Or is it usually _always_ turned OFF, and the Release Manager is not supposed to constantly change it, and I was just unlucky to check today at a coincidental time that was just after you recently decided to temporarily turn it ON just for this sprint?

If tomorrow I'm covering for Release Management of PCAndroid, what am I supposed to do on Fridays then? Am I supposed to submit for review but also turn Managed Publishing off? And then remember to turn in back ON on Monday? Or is it supposed to always be turned OFF and I'm not supposed to touch it (and the fact that it's ON right now is just a temporary glitch)?!

Comment (AliSoftware) [reply fastlane/Fastfile]: Updated the automation to use 1% instead in [2f18528](https://github.com/Automattic/pocket-casts-android/pull/3333/commits/2f1852882c6d3e3113d494ee968c1fe699ceab67) btw; I'll rework the changes I was working on for the MC scenario accordingly.

Comment (MiSikora) [reply fastlane/Fastfile]: > especially given right now it's turned ON

It's not. This is what I see in the Publishing Overview.

Screenshot 2024-12-05 at 20 40 23

I'm not sure how to answer the rest of your questions. I haven't been juggling Managed Publishing. It's off whenever I start the publishing process. I can't guarantee what others have been doing but I'm 99% sure they haven't been juggling it as well. Maybe it happens one in a while like with `7.74` which got released on October 14th but was submitted on October 11th? I can't really say what's been going on.

To reiterate I haven't turned managed publishing on or off during the release. I remember doing it only once when when we needed to coordinate things for one of the features. And If I recall correctly it was done for a beta release.

Comment (AliSoftware) [reply fastlane/Fastfile]: Ooof ok  I think I need to get some rest  🤦 because I was sure I checked and saw it ON… but now I've rechecked and it's indeed OFF for PCAndroid… So I guess I was too tired by my day and probably accidentally checked the GPC page for the wrong app or something when I saw it being ON…  🤦

My bad, and thanks for insisting and clarifying.

If it's always been OFF all that time for PCAndroid that clarifies things a lot, you can ignore my above questions then, sorry 😓

**Comment by AliSoftware** [comment]
## Corresponding changes in the MC scenario

(internal diff ref: D167967-code):

### Moving PRs to next milestone

> Will there still be a step to verify hanging PRs first? Sometimes we want to add changes on the day of the beta release and it would be good to keep this in the MC flow.

Here is how it looks:
image

### Rollout update tasks

No more "Submit the beta version to 20%" task
Instead, the previous task about downloading the 3 APKs and testing them contains this additional detail to remove from review if issues were found during the testing of the 3 APKs (which should be rare anyway)

image

Also, the tasks which previously consisted of going to Google Play to update the rollout of each of the 3 variants manually is now replaced by just these 2 below.

image

_Note that under the "Show Details" of the first task there are fallback instructions to do the rollout update manually in case the lane failed for some reason. The same applied for the similar tasks we had for Intermediate Betas._

### Submission tasks

For the Submission, the 3 tasks about "Verify that the release was uploaded to Play Store as draft" + "repeat for Automotive" + "repeat for Wear" have been removed, and the "Get the Release Notes" + "repeat for Automotive" + "repeat for Wear" + "Submit for Review have all been replaced by this single task:
image

Indeed. the 3 builds will now be submitted for review to 1% by automation right away, and the task about editing them to paste the Release Notes will already act as a way to verify they have been submitted as expected.

### Production Rollout Increase

The tasks about increasing rollout to 10% on Monday (+ repeat for auto + for wear), then the ones to increase to 50% and 100%, have all been replaced by a single task instructing to run `bundle exec fastlane update_rollouts percent:…`. The task about creating a Slack reminder + updating to 50% in 12h have been split in 2 to accomodate:
image

File: fastlane/Fastfile
Comment (mebarbosa) [inline fastlane/Fastfile:158]: This is nice! 👌

**Review by mebarbosa — APPROVED** [review]

File: fastlane/Fastfile
Comment (AliSoftware) [inline fastlane/Fastfile:346]: After discussing with @MiSikora in DM, I'll revert this change of auto-submitting the builds for review—especially because PocketCasts Android does not use Managed Publishing in the Play Store Console.
We'll revert back to submitting them initially as draft to let the Release Manager test the APKs before calling `update_rollouts` only once the tests of the 3 APKs are satisfactory.

**Pull request merged**
